### PR TITLE
Feature/hold my beer

### PR DESCRIPTION
**Pull request by RobHelgeson**
This pr is completely experimental and if it never goes further than a couple of people taking a look, I'm all good with that :)

## What is the current behavior?

`get_highlighted_text` is currently implemented with some nice hand written looping methods. It does however have a small issue with integrating the morph highlighting and html rubies, as outlined in the docs: https://mortii.github.io/anki-morphs/user_guide/known-problems.html under the heading `Ruby characters (furigana, etc.) are displayed wrong in am-highlighted`

## What is the new behavior?

the` get_highlighted_text` process has been completely rewritten to rely more heavily on regexes.

Contrasting the implementations:
* This one takes a lot more bytes to get the job done. -- there are 2 reasons  for this. one, we need to work ruby and span tags together. two, this implementation chooses to tag "unprocessed" sections so the user can style them.
* This one allows correct display of rubies in languages that use them.
* As noted above, this one tags unprocessed sections as unprocessed instead of unknown, as I believe them to be different, and worth differentiating.
* Perf testing has not been comprehensive, but what perf testing I have done indicates that both implementations take roughly the same time.

## What kind of changes does this PR introduce?

- [X] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] This change requires a documentation update

## Checklist:

- [ ] My code successfully passes pre-commit
DOES NOT PASS TESTS THAT USE THE TEST DECKS THEY NEED TO BE REGENERATED AND I DONT KNOW HOW -- I fully anticipate that once regenerated, they'll all pass.
I have written tests for all of the new code, based on the tests for the old code. all of these pass.
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I am willing to help create documentation/guides for the changes made in this PR
- [X] This PR can be rebased onto main without conflicts (create a backup branch before attempting a rebase)
- [ ] I have squashed my commits into sensible portions

### CODE REVIEW

File: test/tests/text_highlighting_test.py [hunk 1]
```diff
@@ -12,9 +12,9 @@
 
 import pytest
 
-from ankimorphs import text_highlighting
 from ankimorphs.ankimorphs_config import AnkiMorphsConfig
 from ankimorphs.morpheme import Morpheme
+from ankimorphs.text_highlighting import get_highlighted_text
```
Comment (RobHelgeson) [inline test/tests/text_highlighting_test.py:17]:
```suggestion
from ankimorphs import text_highlighting
```

File: docs/src/user_guide/setup/settings/extra-fields.md [hunk 2]
```diff
@@ -151,7 +151,12 @@ template, e.g:
...
-You also have to have the `Ignore content in square brackets` [preprocess setting](preprocess.md) activated.
+If you would like to have highlighting, but would prefer to not add this extra field, explore the 
+[am-highlight filter](../../usage/template-filters.md#am-highlight) that can highlight just-in-time, with some 
+tradeoffs.
+
+For either of these options you also have to have the 
+`Ignore content in square brackets` [preprocess setting](preprocess.md) activated.
 
 > **Note**: This does not always work flawlessly. The [known problems](../../known-problems.md) section has more
 details on how to fix ruby character highlighting problems.
```
Comment (RobHelgeson) [inline docs/src/user_guide/setup/settings/extra-fields.md:162]: Self: can remove this.

**Comment by mortii** [comment]
I ran it a couple of times a with a 21k card collection and the speed difference does indeed looks negligible:
 
```text
pr spacy:
Recalc duration: 10.611 seconds
Recalc duration: 10.618 seconds
Recalc duration: 10.574 seconds
Recalc duration: 10.655 seconds
Recalc duration: 10.619 seconds

main spacy:
Recalc duration: 10.831 seconds
Recalc duration: 10.786 seconds
Recalc duration: 10.567 seconds
Recalc duration: 10.706 seconds
Recalc duration: 10.56 seconds

pr mecab:
Recalc duration: 8.435 seconds
Recalc duration: 6.727 seconds
Recalc duration: 6.686 seconds
Recalc duration: 6.749 seconds
Recalc duration: 6.713 seconds

main mecab:
Recalc duration: 6.755 seconds
Recalc duration: 6.862 seconds
Recalc duration: 7.015 seconds
Recalc duration: 6.685 seconds
Recalc duration: 6.705 seconds
```

I don't see any tags for the unprocessed text though. Here is an example text `こォろォさァれェるゥ！` using the mecab (`AnkiMorphs: Japanese`) morphemizer.
 
Screenshot from 2024-12-07 11-49-49

**Comment by mortii** [comment]
Ignore my previous comment, I had the wrong add-on activated on this OS. I'm very sleep deprived.

**Comment by mortii** [comment]
Ok, here is the actual comparison with 21k cards and an `AMD Ryzen™ 7 3700X × 16` :

```text
pr mecab:
Recalc duration: 21.392 seconds
Recalc duration: 19.75 seconds
Recalc duration: 19.958 seconds
Recalc duration: 19.897 seconds
Recalc duration: 19.768 seconds

main mecab:
Recalc duration: 6.755 seconds
Recalc duration: 6.862 seconds
Recalc duration: 7.015 seconds
Recalc duration: 6.685 seconds
Recalc duration: 6.705 seconds

pr spacy:
Recalc duration: 25.834 seconds
Recalc duration: 23.998 seconds
Recalc duration: 24.021 seconds
Recalc duration: 24.03 seconds
Recalc duration: 23.874 seconds

main spacy:
Recalc duration: 10.831 seconds
Recalc duration: 10.786 seconds
Recalc duration: 10.567 seconds
Recalc duration: 10.706 seconds
Recalc duration: 10.56 seconds
```

I think anything above a 20% increase is too slow. 

Regexes can have a pretty disastrous impact on performance :(

**Comment by RobHelgeson** [comment]
> Ok, here is the actual comparison with 21k cards and an `AMD Ryzen™ 7 3700X × 16` :
> 
> ```
> pr mecab:
> Recalc duration: 21.392 seconds
> Recalc duration: 19.75 seconds
> Recalc duration: 19.958 seconds
> Recalc duration: 19.897 seconds
> Recalc duration: 19.768 seconds
> 
> main mecab:
> Recalc duration: 6.755 seconds
> Recalc duration: 6.862 seconds
> Recalc duration: 7.015 seconds
> Recalc duration: 6.685 seconds
> Recalc duration: 6.705 seconds
> 
> pr spacy:
> Recalc duration: 25.834 seconds
> Recalc duration: 23.998 seconds
> Recalc duration: 24.021 seconds
> Recalc duration: 24.03 seconds
> Recalc duration: 23.874 seconds
> 
> main spacy:
> Recalc duration: 10.831 seconds
> Recalc duration: 10.786 seconds
> Recalc duration: 10.567 seconds
> Recalc duration: 10.706 seconds
> Recalc duration: 10.56 seconds
> ```
> 
> I think anything above a 20% increase is too slow.
> 
> Regexes can have a pretty disastrous impact on performance :(
> 
> PS: you can see the recalc duration output if you run anki from a terminal

Bah!! Honestly I'm not too surprised. I took a really remedial method of testing and I bet caching was in play. 

So this impl. is deficient in all ways but one! :) idk if I'm going to give this another crack but I see a couple of options: 

1) See if I can work the ruby fix in your impl.
2) Hybrid approach - take the ideas from here and work them over to not use regexes. ...this will probably end up looking like what you've got :)
3) I don't think this impl is worth tuning, it's just way too far off the perf. things like compiling the regexes and looking for more early loop exits are prob not going to close a 2x-3x slowdown :)

I'm going to close this pr out, I'll tidy up the the other one, since that at least might be useful. working this in my off time has been really fun.

**Pull request closed without merging**